<!-- Copyright 2014 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Identity By State (IBS) for 1000 Genomes Chromosome 22

The following example makes use of [1000
Genomes](https://cloud.google.com/genomics/data/1000-genomes). For more detail
about how this data was loaded into the Google Genomics API, please see [Google
Genomics Public Data](https://cloud.google.com/genomics/data/1000-genomes).

```{r echo=FALSE, eval=FALSE}
# This codelab assumes that the current working directory is where the Rmd file
# resides
setwd("/YOUR/PATH/TO//1000Genomes-chr22/ibs")
```

## Computing IBS

Follow the instructions at [README.rst](../../../README.rst) to compute the
Identity By State (IBS) scores for chr 22 of the [1000 Genomes data
set](https://cloud.google.com/genomics/data/1000-genomes). Specifically, first
configure and run the IBS DataFlow pipeline:

```sh
    java -cp ../../../target/google-genomics-dataflow-v1beta2-0.2-SNAPSHOT.jar \
    com.google.cloud.genomics.dataflow.pipelines.IdentityByState \
    --runner=BlockingDataflowPipelineRunner \
    --project=my-project-id \
    --stagingLocation=gs://my-bucket/staging \
    --output=gs://my-bucket/output/ibs.tsv \
    --numWorkers=40 \
    --references=22:16050407:51243297
```

We used the following BigQuery command to compute the reference bounds in the
above command:

```sql
SELECT
  MIN(start) AS start,
  MAX(end) AS end
FROM
  [genomics-public-data:1000_genomes.variants]
WHERE
  reference_name IN ("22")
  ```

Next, merge the generated
[`1000genomes_chr22_ibs.tsv`](1000genomes_chr22_ibs.tsv) shards into a single
file:

```sh
    gsutil cat gs://my-bucket/output/is.tsv* | sort > 1000genomes_chr22_ibs.tsv
```

Finally, run [`generate.R`](generate.R) to visualize
[`1000genomes_chr22_ibs.tsv`](1000genomes_chr22_ibs.tsv) and compare it with
[the IBS scores computed by
PLINK/SEQ](https://raw.githubusercontent.com/deflaux/codelabs/qc-codelab/R/1000Genomes-BRCA1-analysis/data/plinkseqIBS/chr22/ALL.chr22.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.ibs).

## Loading IBS Data in R

`ibsFilename1` contains an N^2 x 3 IBS matrix, where N is the size of the
population and each row represents the IBS score for a pair of individuals.

```{r message=FALSE, comment=NA}
ibsFilename1="1000genomes_chr22_ibs.tsv"
```

`ibsFilename2` contains an NxN IBS matrix, where N is the size of the population
and each cell of the matrix contains the IBS score of a pair of individuals.

```{r message=FALSE, comment=NA}
ibsFilename2="1000genomes_phase1_chr22_plinkseq_ibs.tsv"
```

```{r message=FALSE, comment=NA}
library(reshape2)
ReadIBSFile <- function(ibsFilename, header=FALSE, rowNames=NULL) {
  ibsData <- read.table(file=ibsFilename, header=header,
                        row.names=rowNames, stringsAsFactors=FALSE)
  return (ibsData)
}
ibsData1 <- ReadIBSFile(ibsFilename1)
ibsData2 <- ReadIBSFile(ibsFilename2, header=TRUE, rowNames=1)
```

Transform the NxN matrix into a N^2 x 3 matrix, where each row represents the
IBS score for a pair of individuals.

```{r message=FALSE, comment=NA}
ibsData2 <- melt(data.matrix(ibsData2))
```

Set the column names of the two sets of IBS data consistently.

```{r message=FALSE, comment=NA}
ColumnNames <- function(ibsData) {
  colnames(ibsData) <- c("sample1", "sample2", "ibsScore")
}
colnames(ibsData1) <- ColumnNames(ibsData1)
colnames(ibsData2) <- ColumnNames(ibsData2)
```

Make the IBS matrix symmetric.

```{r message=FALSE, comment=NA}
MakeIBSDataSymmetric <- function(ibsData) {
  ibsPairsMirrored <- data.frame(sample1=ibsData$sample2,
                                 sample2=ibsData$sample1,
                                 ibsScore=ibsData$ibsScore)
  ibsData <- rbind(ibsData, ibsPairsMirrored)
}
ibsData1 <- MakeIBSDataSymmetric(ibsData1)
```

## IBS Heat Map

Exclude the IBS values for a genome and itself, because those values are always
1.0 and skew the heat map.

```{r message=FALSE, comment=NA}
ExcludeDiagonal <- function(ibsData) {
  ibsData <- subset(ibsData, ibsData$sample1 != ibsData$sample2)
  return (ibsData)
}
ibsData2Sample <- ExcludeDiagonal(ibsData2)
```

Extract the IBS matrix for a random sample of the individuals.

```{r message=FALSE, comment=NA}
SampleIBSMatrix <- function(ibsData) {
  individuals <- unique(ibsData$sample1)
  sampleSize <- 50
  sample <- sample(individuals, sampleSize)
  ibsData <- subset(ibsData, ibsData$sample1 %in% sample)
  ibsData <- subset(ibsData, ibsData$sample2 %in% sample)
  return (ibsData)
}
ibsData2Sample <- SampleIBSMatrix(ibsData2Sample)
```

Draw a heat map based on the sampled IBS scores.

```{r ibs-heat-map, fig.align="center", fig.width=10, fig.height=10, message=FALSE, comment=NA}
require(ggplot2)
DrawHeatMap <- function(ibsData) {
  p <- ggplot(data=ibsData, aes(x=sample1, y=sample2)) +
       theme(axis.ticks=element_blank(), axis.text=element_blank()) +
       geom_tile(aes(fill=ibsScore), colour="white") +
       scale_fill_gradient(low="white", high="steelblue", na.value="black",
                           guide=guide_colourbar(title= "IBS Score")) +
       labs(list(title="Identity By State (IBS) Heat Map",
                 x="Sample", y="Sample"))
  p
}
DrawHeatMap(ibsData2Sample)
```

## Comparison with [PLINK/SEQ](https://atgu.mgh.harvard.edu/plinkseq/) IBS

Compare the IBS matrix generated by DataFlow with [the one generated by
PLINK/SEQ](https://raw.githubusercontent.com/deflaux/codelabs/qc-codelab/R/1000Genomes-BRCA1-analysis/data/plinkseqIBS/chr22/ALL.chr22.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.ibs)
and report the number of differences.

```{r message=FALSE, comment=NA}
mergedIBS <- merge(ibsData1, ibsData2, by = c(colnames(ibsData1)[1:2]))
diffIBS <- abs(mergedIBS$ibsScore.x - mergedIBS$ibsScore.y) < 1e-6
mergedDiffIBS <- cbind(mergedIBS, diffIBS)
colnames(mergedDiffIBS) <- c(colnames(mergedDiffIBS)[1:4], "almostEqualIBS")
nrow(mergedDiffIBS[mergedDiffIBS$almostEqualIBS == FALSE,])
```

Plot the two IBS matrices to show their linear relationship.

```{r ibs-lm, fig.align="center", fig.width=5, fig.height=5, message=FALSE, comment=NA}
ggplot(mergedIBS, aes(x=ibsScore.x, y=ibsScore.y)) + geom_point()
```

```{r message=FALSE, comment=NA}
model <- lm(ibsScore.y ~ ibsScore.x, mergedIBS)
summary(model)
```

Compare these results to the pedigree.
```{r message=FALSE, comment=NA}
pedigree <- read.delim("ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_g1k.ped", as.is=TRUE)
summary(pedigree)

require(dplyr)
ibs <- filter(ibsData1, sample1 != sample2)
ibs_fam <- inner_join(ibs, select(pedigree, Individual.ID, Family.ID), by=c("sample1" = "Individual.ID"))
ibs_fams <- inner_join(ibs_fam, select(pedigree, Individual.ID, Family.ID), by=c("sample2" = "Individual.ID"))
```
```{r ibs-boxplot, fig.align="center", fig.width=10, fig.height=10, message=FALSE, comment=NA}
boxplot(ibsScore~Family.ID.x == Family.ID.y, data=ibs_fams, main="Identity By State Results compared to Pedigree", 
    xlab="Individuals in Same Family", ylab="IBS Score")
```
